tls/extensions/s2n_client_server_name.c

```diff
@@ -58,26 +58,29 @@ static int s2n_client_server_name_send(struct s2n_connection *conn, struct s2n_s
     return S2N_SUCCESS;
 }
 
-/* Read the extension up to the first item in ServerNameList. Store the first entry's length in server_name_len.
- * For now s2n ignores all subsequent items in ServerNameList.
+/* Read the extension up to the first item in ServerNameList. Instantiates the server_name blob to
+ * point to the first entry. For now s2n ignores all subsequent items in ServerNameList.
  */
-static int s2n_client_server_name_check(struct s2n_connection *conn, struct s2n_stuffer *extension, uint16_t *server_name_len)
+S2N_RESULT s2n_parse_server_name(struct s2n_stuffer *extension, struct s2n_blob *server_name)
 {
-    POSIX_ENSURE_REF(conn);
+    uint16_t list_size = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(extension, &list_size));
+    RESULT_ENSURE_LTE(list_size, s2n_stuffer_data_available(extension));
 
-    uint16_t size_of_all;
-    POSIX_GUARD(s2n_stuffer_read_uint16(extension, &size_of_all));
-    POSIX_ENSURE_LTE(size_of_all, s2n_stuffer_data_available(extension));
+    uint8_t server_name_type = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(extension, &server_name_type));
+    RESULT_ENSURE_EQ(server_name_type, S2N_NAME_TYPE_HOST_NAME);
 
-    uint8_t server_name_type;
-    POSIX_GUARD(s2n_stuffer_read_uint8(extension, &server_name_type));
-    POSIX_ENSURE_EQ(server_name_type, S2N_NAME_TYPE_HOST_NAME);
+    uint16_t length = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(extension, &length));
+    RESULT_ENSURE_LTE(length, s2n_stuffer_data_available(extension));
+    RESULT_ENSURE_LTE(length, S2N_MAX_SERVER_NAME);
 
```

Comment:
I wonder if we want to move this check to s2n_client_server_name_recv instead. I was initially thinking "the two methods should behave exactly the same", but technically the RFC allows larger server names, and the client hello method doesn't have the same restriction on storage space. I don't feel strongly either way-- we can always relax the check later.

Reply:
I moved it to s2n_client_server_name_recv; couldn't really think of a good reason to keep the check in both.